File: src/engine.rs
use crate::memory::ResultArena;
use crate::diff::compute_compact_diff_v2;
use crate::error::{ErrorBuffer, EngineError};
use crate::parser::CompactParser;
use crate::status::Status;
use crate::config::EngineConfig;

pub struct Engine {
    magic: u32,
    arena: ResultArena,
    left_parser: CompactParser,
    right_parser: CompactParser,
    error: ErrorBuffer,
    sealed: bool,
    left_input: Vec<u8>,
    right_input: Vec<u8>,
    left_index: crate::simd_index::StructuralIndex,
    right_index: crate::simd_index::StructuralIndex,
}

impl Engine {
    pub fn new(config: EngineConfig, magic: u32) -> Result<Self, EngineError> {
        let input_cap = (config.max_input_size / 2) as usize;
        Ok(Self {
            magic,
            arena: ResultArena::new(config.max_memory_bytes),
            left_parser: CompactParser::new(config.max_object_keys, config.compute_mode),
            right_parser: CompactParser::new(config.max_object_keys, config.compute_mode),
            error: ErrorBuffer::new(),
            sealed: false,
            left_input: Vec::with_capacity(input_cap),
            right_input: Vec::with_capacity(input_cap),
            left_index: crate::simd_index::StructuralIndex::new(),
            right_index: crate::simd_index::StructuralIndex::new(),
        })
    }

    pub fn magic(&self) -> u32 { self.magic }
    pub fn clear_magic(&mut self) { self.magic = 0; }

    pub fn commit_left(&mut self, len: u32) -> Status {
        if self.sealed { return Status::EngineSealed; }
        // `len` is supplied by the host. Reading past the buffer's allocated
        // capacity would be out-of-bounds — reject instead of trusting it.
        if len as usize > self.left_input.capacity() { return Status::Error; }
        let bytes = unsafe { std::slice::from_raw_parts(self.left_input.as_ptr(), len as usize) };
        self.left_index.build(bytes);
        match self.left_parser.parse_with_index(bytes, &self.left_index) {
            Ok(_) => Status::Ok,
            Err(_) => Status::Error,
        }
    }

    pub fn commit_right(&mut self, len: u32) -> Status {
        if self.sealed { return Status::EngineSealed; }
        if len as usize > self.right_input.capacity() { return Status::Error; }
        let bytes = unsafe { std::slice::from_raw_parts(self.right_input.as_ptr(), len as usize) };
        self.right_index.build(bytes);
        match self.right_parser.parse_with_index(bytes, &self.right_index) {
            Ok(_) => Status::Ok,
            Err(_) => Status::Error,
        }
    }

    pub fn finalize(&mut self) -> Result<*const u8, EngineError> {
        if self.sealed { return Ok(self.arena.as_ptr()); }
        self.sealed = true;

        let diffs = compute_compact_diff_v2(&self.left_parser, &self.right_parser);
        
        for d in diffs {
            if let Err(_) = self.arena.write_entry_v2(
                d.op,
                d.path_id,
                d.left_val,
                d.right_val,
            ) {
                self.error.set(&EngineError::MemoryLimitExceeded);
                break;
            }
        }

        self.arena.seal();
        Ok(self.arena.as_ptr())
    }

    pub fn clear(&mut self) {
        self.arena.clear();
        self.left_parser.clear();
        self.right_parser.clear();
        self.sealed = false;
    }

    pub fn left_input_ptr(&mut self) -> *mut u8 { self.left_input.as_mut_ptr() }
    pub fn right_input_ptr(&mut self) -> *mut u8 { self.right_input.as_mut_ptr() }

    pub fn result_len(&self) -> u32 { self.arena.len() }
    pub fn last_error_len(&self) -> u32 { self.error.len() }
    pub fn last_error_ptr(&self) -> *const u8 { self.error.as_ptr() }
}

File: cpp/src/engine.cpp
#include "parser.h"
#include "path.h"
#include "diff.h"
#include <vector>
#include <cstring>

namespace diffcore {

class DiffEngine {
public:
    DiffEngine(uint32_t max_memory, uint32_t max_input)
        : max_input_(max_input) {
        left_input_.resize(max_input / 2);
        right_input_.resize(max_input / 2);
        result_buffer_.reserve(16384);
    }
    
    uint8_t* left_input_ptr() { return left_input_.data(); }
    uint8_t* right_input_ptr() { return right_input_.data(); }
    
    int32_t commit_left(uint32_t len) {
        if (len > left_input_.size()) return -1;
        return left_parser_.parse(left_input_.data(), len, path_arena_) ? 0 : -1;
    }
    
    int32_t commit_right(uint32_t len) {
        if (len > right_input_.size()) return -1;
        return right_parser_.parse(right_input_.data(), len, path_arena_) ? 0 : -1;
    }
    
    const uint8_t* finalize() {
        diffs_ = compute_diff(left_parser_, right_parser_);
        
        // Build result buffer (binary format)
        result_buffer_.clear();
        
        // Header: [u16 major][u16 minor][u32 count][u64 total_len]
        uint16_t major = 2, minor = 1;
        result_buffer_.insert(result_buffer_.end(), reinterpret_cast<uint8_t*>(&major), 
                             reinterpret_cast<uint8_t*>(&major) + 2);
        result_buffer_.insert(result_buffer_.end(), reinterpret_cast<uint8_t*>(&minor), 
                             reinterpret_cast<uint8_t*>(&minor) + 2);
        uint32_t count = static_cast<uint32_t>(diffs_.size());
        result_buffer_.insert(result_buffer_.end(), reinterpret_cast<uint8_t*>(&count), 
                             reinterpret_cast<uint8_t*>(&count) + 4);
        uint64_t placeholder = 0;
        result_buffer_.insert(result_buffer_.end(), reinterpret_cast<uint8_t*>(&placeholder), 
                             reinterpret_cast<uint8_t*>(&placeholder) + 8);
        
        // Entries (24 bytes each)
        for (const auto& d : diffs_) {
            result_buffer_.push_back(static_cast<uint8_t>(d.op));
            result_buffer_.insert(result_buffer_.end(), reinterpret_cast<const uint8_t*>(&d.path_id),
                                 reinterpret_cast<const uint8_t*>(&d.path_id) + 4);
            result_buffer_.insert(result_buffer_.end(), reinterpret_cast<const uint8_t*>(&d.left_offset),
                                 reinterpret_cast<const uint8_t*>(&d.left_offset) + 4);
            result_buffer_.insert(result_buffer_.end(), reinterpret_cast<const uint8_t*>(&d.left_len),
                                 reinterpret_cast<const uint8_t*>(&d.left_len) + 4);
            result_buffer_.insert(result_buffer_.end(), reinterpret_cast<const uint8_t*>(&d.right_offset),
                                 reinterpret_cast<const uint8_t*>(&d.right_offset) + 4);
            result_buffer_.insert(result_buffer_.end(), reinterpret_cast<const uint8_t*>(&d.right_len),
                                 reinterpret_cast<const uint8_t*>(&d.right_len) + 4);
            // Padding
            result_buffer_.push_back(0);
            result_buffer_.push_back(0);
            result_buffer_.push_back(0);
        }
        
        // Update total len in header
        uint64_t total_len = result_buffer_.size();
        std::memcpy(result_buffer_.data() + 8, &total_len, 8);
        
        return result_buffer_.data();
    }
    
    uint32_t result_len() const { return static_cast<uint32_t>(result_buffer_.size()); }
    
    const uint8_t* batch_resolve_symbols(uint32_t* out_len) {
        symbol_buffer_.clear();
        uint32_t count = static_cast<uint32_t>(diffs_.size());
        symbol_buffer_.insert(symbol_buffer_.end(), reinterpret_cast<uint8_t*>(&count),
                             reinterpret_cast<uint8_t*>(&count) + 4);
        
        for (const auto& d : diffs_) {
            std::string path = path_arena_.path_to_string(d.path_id);
            uint32_t len = static_cast<uint32_t>(path.size());
            symbol_buffer_.insert(symbol_buffer_.end(), reinterpret_cast<uint8_t*>(&len),
                                 reinterpret_cast<uint8_t*>(&len) + 4);
            symbol_buffer_.insert(symbol_buffer_.end(), path.begin(), path.end());
        }
        
        *out_len = static_cast<uint32_t>(symbol_buffer_.size());
        return symbol_buffer_.data();
    }
    
    void clear() {
        left_parser_.clear();
        right_parser_.clear();
        path_arena_.clear();
        diffs_.clear();
        result_buffer_.clear();
        symbol_buffer_.clear();
    }

private:
    CompactParser left_parser_;
    CompactParser right_parser_;
    PathArena path_arena_;
    std::vector<DiffEntry> diffs_;
    std::vector<uint8_t> left_input_;
    std::vector<uint8_t> right_input_;
    std::vector<uint8_t> result_buffer_;
    std::vector<uint8_t> symbol_buffer_;
    uint32_t max_input_;
};

} // namespace diffcore
